File: building_block_apps/spark/python/spark_common.py
#!/usr/bin/env python3
"""
Common utilities and configuration for Spark building block applications.

This module provides:
- Spark session creation with optimized settings
- Common data schemas and sample data generation
- Utility functions for performance monitoring
- Error handling patterns
"""

import sys
import time
import json
import random
import argparse
from typing import Dict, Any, List, Optional
from dataclasses import dataclass
from datetime import datetime, timedelta

try:
    from pyspark.sql import SparkSession, DataFrame
    from pyspark.sql.types import (
        StructType, StructField, StringType, IntegerType, 
        DoubleType, TimestampType, BooleanType
    )
    from pyspark.sql.functions import col, current_timestamp, lit
except ImportError:
    print("ERROR: PySpark not found. Install with: pip install pyspark")
    sys.exit(1)


@dataclass
class SparkConfig:
    """Configuration for Spark applications."""
    app_name: str = "SparkBuildingBlock"
    master: str = "spark://192.168.1.184:7077"
    executor_memory: str = "2g"
    executor_cores: str = "2"
    driver_memory: str = "1g"
    kafka_servers: str = "192.168.1.184:9092,192.168.1.187:9092,192.168.1.190:9092"
    checkpoint_dir: str = "/tmp/spark-checkpoints"
    warehouse_dir: str = "/home/spark/spark/warehouse"


def create_spark_session(config: SparkConfig) -> SparkSession:
    """
    Create a Spark session with optimized configuration.
    
    Args:
        config: SparkConfig object with application settings
        
    Returns:
        Configured SparkSession
    """
    builder = SparkSession.builder \
        .appName(config.app_name) \
        .master(config.master) \
        .config("spark.executor.memory", config.executor_memory) \
        .config("spark.executor.cores", config.executor_cores) \
        .config("spark.driver.memory", config.driver_memory) \
        .config("spark.sql.warehouse.dir", config.warehouse_dir) \
        .config("spark.serializer", "org.apache.spark.serializer.KryoSerializer") \
        .config("spark.sql.adaptive.enabled", "true") \
        .config("spark.sql.adaptive.coalescePartitions.enabled", "true") \
        .config("spark.sql.adaptive.skewJoin.enabled", "true") \
        .config("spark.dynamicAllocation.enabled", "false")
    
    # Add Kafka dependencies if available
    try:
        builder = builder.config(
            "spark.jars.packages", 
            "org.apache.spark:spark-sql-kafka-0-10_2.12:3.5.6"
        )
    except Exception:
        print("WARNING: Kafka package not available, streaming examples may fail")
    
    session = builder.getOrCreate()
    session.sparkContext.setLogLevel("WARN")
    
    print(f"✅ Spark session created: {config.app_name}")
    print(f"   Master: {config.master}")
    print(f"   Executor Memory: {config.executor_memory}")
    print(f"   Executor Cores: {config.executor_cores}")
    
    return session


def get_sample_data_schema() -> StructType:
    """
    Define a common schema for sample data used across examples.
    
    Returns:
        StructType schema for sample e-commerce data
    """
    return StructType([
        StructField("user_id", StringType(), True),
        StructField("product_id", StringType(), True),
        StructField("category", StringType(), True),
        StructField("action", StringType(), True),
        StructField("timestamp", TimestampType(), True),
        StructField("price", DoubleType(), True),
        StructField("quantity", IntegerType(), True),
        StructField("session_id", StringType(), True),
        StructField("is_premium", BooleanType(), True),
        StructField("region", StringType(), True),
    ])


def generate_sample_data(spark: SparkSession, num_records: int = 10000) -> DataFrame:
    """
    Generate sample e-commerce data for testing and examples.
    
    Args:
        spark: SparkSession 
        num_records: Number of records to generate
        
    Returns:
        DataFrame with sample data
    """
    print(f"🔄 Generating {num_records:,} sample records...")
    
    categories = ["electronics", "clothing", "books", "home", "sports"]
    actions = ["view", "add_to_cart", "purchase", "remove_from_cart"]
    regions = ["US-East", "US-West", "EU", "Asia"]
    
    # Generate data in batches for better performance
    batch_size = 1000
    data = []
    
    for i in range(num_records):
        record = {
            "user_id": f"user_{random.randint(1, 50000)}",
            "product_id": f"prod_{random.randint(1, 10000)}",
            "category": random.choice(categories),
            "action": random.choice(actions),
            "timestamp": datetime.now() - timedelta(
                seconds=random.randint(0, 7*24*3600)  # Last 7 days
            ),
            "price": round(random.uniform(10, 1000), 2),
            "quantity": random.randint(1, 5),
            "session_id": f"session_{random.randint(1, 100000)}",
            "is_premium": random.choice([True, False]),
            "region": random.choice(regions),
        }
        data.append(record)
        
        if len(data) >= batch_size:
            # Process batch
            if i % 5000 == 0 and i > 0:
                print(f"   Generated {i:,} records...")
    
    df = spark.createDataFrame(data, schema=get_sample_data_schema())
    print(f"✅ Generated {df.count():,} records")
    return df


def monitor_job_progress(spark: SparkSession, job_name: str = "Spark Job"):
    """
    Print job progress and performance metrics.
    
    Args:
        spark: SparkSession
        job_name: Name of the job for logging
    """
    print(f"\n📊 {job_name} Metrics:")
    print(f"   Spark UI: http://192.168.1.184:4040")
    
    # Get basic stats from Spark context
    sc = spark.sparkContext
    print(f"   Application ID: {sc.applicationId}")
    print(f"   Default Parallelism: {sc.defaultParallelism}")
    
    # Additional metrics if available
    try:
        status = sc.statusTracker()
        executor_infos = status.getExecutorInfos()
        print(f"   Active Executors: {len(executor_infos)}")
        
        total_cores = sum(exec.totalCores for exec in executor_infos)
        print(f"   Total Cores: {total_cores}")
        
    except Exception as e:
        print(f"   Metrics unavailable: {e}")


def setup_kafka_checkpoint(spark: SparkSession, config: SparkConfig, app_name: str):
    """
    Setup checkpoint directory for streaming applications.
    
    Args:
        spark: SparkSession
        config: SparkConfig
        app_name: Application name for unique checkpoint directory
    """
    import os
    checkpoint_path = f"{config.checkpoint_dir}/{app_name}"
    
    # Create checkpoint directory if it doesn't exist
    try:
        os.makedirs(checkpoint_path, exist_ok=True)
        print(f"✅ Checkpoint directory: {checkpoint_path}")
    except Exception as e:
        print(f"⚠️  Could not create checkpoint directory: {e}")
        print(f"   Using default: /tmp/spark-checkpoints/{app_name}")


def parse_common_args() -> argparse.ArgumentParser:
    """
    Create argument parser with common Spark application options.
    
    Returns:
        ArgumentParser with common options
    """
    parser = argparse.ArgumentParser(description="Spark Building Block Application")
    
    parser.add_argument(
        "--master", 
        default="spark://192.168.1.184:7077",
        help="Spark master URL (default: spark://192.168.1.184:7077)"
    )
    
    parser.add_argument(
        "--app-name",
        default="SparkBuildingBlock", 
        help="Spark application name"
    )
    
    parser.add_argument(
        "--executor-memory",
        default="2g",
        help="Executor memory (default: 2g)"
    )
    
    parser.add_argument(
        "--executor-cores", 
        default="2",
        help="Executor cores (default: 2)"
    )
    
    parser.add_argument(
        "--records",
        type=int,
        default=10000,
        help="Number of records to generate/process (default: 10000)"
    )
    
    parser.add_argument(
        "--output-path",
        default="/tmp/spark-output",
        help="Output path for results (default: /tmp/spark-output)"
    )
    
    parser.add_argument(
        "--kafka-topic",
        default="spark-demo",
        help="Kafka topic for streaming examples (default: spark-demo)"
    )
    
    parser.add_argument(
        "--duration",
        type=int,
        default=60,
        help="Duration to run streaming jobs in seconds (default: 60)"
    )
    
    parser.add_argument(
        "--verbose", "-v",
        action="store_true",
        help="Enable verbose logging"
    )
    
    return parser


def cleanup_session(spark: SparkSession):
    """
    Properly clean up Spark session.
    
    Args:
        spark: SparkSession to clean up
    """
    try:
        print("\n🔄 Cleaning up Spark session...")
        spark.stop()
        print("✅ Spark session stopped")
    except Exception as e:
        print(f"⚠️  Error during cleanup: {e}")


def print_dataframe_summary(df: DataFrame, name: str = "DataFrame", show_data: bool = True):
    """
    Print a comprehensive summary of a DataFrame.
    
    Args:
        df: DataFrame to summarize
        name: Name for the DataFrame in output
        show_data: Whether to show sample data
    """
    print(f"\n📊 {name} Summary:")
    print(f"   Rows: {df.count():,}")
    print(f"   Columns: {len(df.columns)}")
    print(f"   Schema:")
    
    for field in df.schema.fields:
        print(f"     {field.name}: {field.dataType}")
    
    if show_data and df.count() > 0:
        print(f"\n   Sample Data:")
        df.show(5, truncate=False)
        
        # Show basic statistics for numeric columns
        numeric_cols = [f.name for f in df.schema.fields 
                       if isinstance(f.dataType, (IntegerType, DoubleType))]
        if numeric_cols:
            print(f"\n   Numeric Statistics:")
            df.select(*numeric_cols).describe().show()


def handle_errors(func):
    """
    Decorator for consistent error handling across Spark applications.
    
    Args:
        func: Function to wrap with error handling
    """
    def wrapper(*args, **kwargs):
        try:
            return func(*args, **kwargs)
        except KeyboardInterrupt:
            print("\n⚠️  Application interrupted by user")
            sys.exit(0)
        except Exception as e:
            print(f"\n❌ Error in {func.__name__}: {e}")
            print(f"   Error type: {type(e).__name__}")
            import traceback
            traceback.print_exc()
            sys.exit(1)
    return wrapper


if __name__ == "__main__":
    # Test the common utilities
    print("🧪 Testing Spark common utilities...")
    
    config = SparkConfig(app_name="SparkCommonTest")
    spark = create_spark_session(config)
    
    # Generate and show sample data
    df = generate_sample_data(spark, 100)
    print_dataframe_summary(df, "Sample Data")
    
    monitor_job_progress(spark, "Common Utilities Test")
    cleanup_session(spark)
    
    print("✅ Spark common utilities test completed")

File: building_block_apps/spark/java/SparkCommon.java
import org.apache.spark.SparkConf;
import org.apache.spark.sql.SparkSession;
import org.apache.spark.sql.Dataset;
import org.apache.spark.sql.Row;
import org.apache.spark.sql.types.StructType;
import org.apache.spark.sql.types.DataTypes;
import org.apache.spark.sql.types.StructField;
import org.apache.spark.sql.functions;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.Serializable;
import java.sql.Timestamp;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.*;
import java.util.concurrent.ThreadLocalRandom;

/**
 * Common utilities and configuration for Spark building block applications in Java.
 * 
 * This class provides:
 * - Spark session creation with optimized settings
 * - Common data schemas and sample data generation
 * - Utility functions for performance monitoring
 * - Error handling patterns
 * - Configuration management
 */
public class SparkCommon implements Serializable {
    
    private static final Logger logger = LoggerFactory.getLogger(SparkCommon.class);
    private static final String DEFAULT_MASTER = "spark://192.168.1.184:7077";
    private static final String DEFAULT_APP_NAME = "SparkBuildingBlock";
    
    /**
     * Configuration class for Spark applications.
     */
    public static class SparkConfig implements Serializable {
        private String appName;
        private String master;
        private String executorMemory;
        private String executorCores;
        private String driverMemory;
        private String kafkaServers;
        private String checkpointDir;
        private String warehouseDir;
        
        public SparkConfig() {
            this.appName = DEFAULT_APP_NAME;
            this.master = DEFAULT_MASTER;
            this.executorMemory = "2g";
            this.executorCores = "2";
            this.driverMemory = "1g";
            this.kafkaServers = "192.168.1.184:9092,192.168.1.187:9092,192.168.1.190:9092";
            this.checkpointDir = "/tmp/spark-checkpoints";
            this.warehouseDir = "/home/spark/spark/warehouse";
        }
        
        // Getters and setters
        public String getAppName() { return appName; }
        public void setAppName(String appName) { this.appName = appName; }
        
        public String getMaster() { return master; }
        public void setMaster(String master) { this.master = master; }
        
        public String getExecutorMemory() { return executorMemory; }
        public void setExecutorMemory(String executorMemory) { this.executorMemory = executorMemory; }
        
        public String getExecutorCores() { return executorCores; }
        public void setExecutorCores(String executorCores) { this.executorCores = executorCores; }
        
        public String getDriverMemory() { return driverMemory; }
        public void setDriverMemory(String driverMemory) { this.driverMemory = driverMemory; }
        
        public String getKafkaServers() { return kafkaServers; }
        public void setKafkaServers(String kafkaServers) { this.kafkaServers = kafkaServers; }
        
        public String getCheckpointDir() { return checkpointDir; }
        public void setCheckpointDir(String checkpointDir) { this.checkpointDir = checkpointDir; }
        
        public String getWarehouseDir() { return warehouseDir; }
        public void setWarehouseDir(String warehouseDir) { this.warehouseDir = warehouseDir; }
    }
    
    /**
     * Create a Spark session with optimized configuration.
     * 
     * @param config SparkConfig object with application settings
     * @return Configured SparkSession
     */
    public static SparkSession createSparkSession(SparkConfig config) {
        logger.info("Creating Spark session: {}", config.getAppName());
        
        SparkConf sparkConf = new SparkConf()
            .setAppName(config.getAppName())
            .setMaster(config.getMaster())
            .set("spark.executor.memory", config.getExecutorMemory())
            .set("spark.executor.cores", config.getExecutorCores())
            .set("spark.driver.memory", config.getDriverMemory())
            .set("spark.sql.warehouse.dir", config.getWarehouseDir())
            .set("spark.serializer", "org.apache.spark.serializer.KryoSerializer")
            .set("spark.sql.adaptive.enabled", "true")
            .set("spark.sql.adaptive.coalescePartitions.enabled", "true")
            .set("spark.sql.adaptive.skewJoin.enabled", "true")
            .set("spark.dynamicAllocation.enabled", "false");
        
        // Add Kafka dependencies if available
        try {
            sparkConf.set("spark.jars.packages", "org.apache.spark:spark-sql-kafka-0-10_2.12:3.5.6");
        } catch (Exception e) {
            logger.warn("Kafka package not available, streaming examples may fail: {}", e.getMessage());
        }
        
        SparkSession spark = SparkSession.builder()
            .config(sparkConf)
            .getOrCreate();
        
        spark.sparkContext().setLogLevel("WARN");
        
        logger.info("✅ Spark session created successfully");
        logger.info("   Master: {}", config.getMaster());
        logger.info("   Executor Memory: {}", config.getExecutorMemory());
        logger.info("   Executor Cores: {}", config.getExecutorCores());
        
        return spark;
    }
    
    /**
     * Define a common schema for sample data used across examples.
     * 
     * @return StructType schema for sample e-commerce data
     */
    public static StructType getSampleDataSchema() {
        return DataTypes.createStructType(new StructField[] {
            DataTypes.createStructField("user_id", DataTypes.StringType, true),
            DataTypes.createStructField("product_id", DataTypes.StringType, true),
            DataTypes.createStructField("category", DataTypes.StringType, true),
            DataTypes.createStructField("action", DataTypes.StringType, true),
            DataTypes.createStructField("timestamp", DataTypes.TimestampType, true),
            DataTypes.createStructField("price", DataTypes.DoubleType, true),
            DataTypes.createStructField("quantity", DataTypes.IntegerType, true),
            DataTypes.createStructField("session_id", DataTypes.StringType, true),
            DataTypes.createStructField("is_premium", DataTypes.BooleanType, true),
            DataTypes.createStructField("region", DataTypes.StringType, true)
        });
    }
    
    /**
     * Generate sample e-commerce data for testing and examples.
     * 
     * @param spark SparkSession
     * @param numRecords Number of records to generate
     * @return Dataset with sample data
     */
    public static Dataset<Row> generateSampleData(SparkSession spark, int numRecords) {
        logger.info("🔄 Generating {} sample records...", numRecords);
        
        String[] categories = {"electronics", "clothing", "books", "home", "sports"};
        String[] actions = {"view", "add_to_cart", "purchase", "remove_from_cart"};
        String[] regions = {"US-East", "US-West", "EU", "Asia"};
        
        List<Row> data = new ArrayList<>();
        Random random = new Random();
        
        for (int i = 0; i < numRecords; i++) {
            LocalDateTime timestamp = LocalDateTime.now().minusSeconds(random.nextInt(7 * 24 * 3600));
            
            Row record = org.apache.spark.sql.RowFactory.create(
                "user_" + (random.nextInt(50000) + 1),
                "prod_" + (random.nextInt(10000) + 1),
                categories[random.nextInt(categories.length)],
                actions[random.nextInt(actions.length)],
                Timestamp.valueOf(timestamp),
                Math.round((10 + random.nextDouble() * 990) * 100.0) / 100.0,
                random.nextInt(5) + 1,
                "session_" + (random.nextInt(100000) + 1),
                random.nextBoolean(),
                regions[random.nextInt(regions.length)]
            );
            data.add(record);
            
            if ((i + 1) % 5000 == 0) {
                logger.info("   Generated {} records...", i + 1);
            }
        }
        
        Dataset<Row> df = spark.createDataFrame(data, getSampleDataSchema());
        long count = df.count();
        logger.info("✅ Generated {} records", count);
        
        return df;
    }
    
    /**
     * Print job progress and performance metrics.
     * 
     * @param spark SparkSession
     * @param jobName Name of the job for logging
     */
    public static void monitorJobProgress(SparkSession spark, String jobName) {
        logger.info("\n📊 {} Metrics:", jobName);
        logger.info("   Spark UI: http://192.168.1.184:4040");
        
        try {
            String applicationId = spark.sparkContext().applicationId();
            int defaultParallelism = spark.sparkContext().defaultParallelism();
            
            logger.info("   Application ID: {}", applicationId);
            logger.info("   Default Parallelism: {}", defaultParallelism);
            
        } catch (Exception e) {
            logger.warn("   Metrics unavailable: {}", e.getMessage());
        }
    }
    
    /**
     * Print a comprehensive summary of a Dataset.
     * 
     * @param df Dataset to summarize
     * @param name Name for the Dataset in output
     * @param showData Whether to show sample data
     */
    public static void printDatasetSummary(Dataset<Row> df, String name, boolean showData) {
        logger.info("\n📊 {} Summary:", name);
        
        try {
            long count = df.count();
            int columnCount = df.columns().length;
            
            logger.info("   Rows: {}", count);
            logger.info("   Columns: {}", columnCount);
            logger.info("   Schema:");
            
            StructType schema = df.schema();
            for (StructField field : schema.fields()) {
                logger.info("     {}: {}", field.name(), field.dataType().simpleString());
            }
            
            if (showData && count > 0) {
                logger.info("\n   Sample Data:");
                df.show(5, false);
                
                // Show basic statistics for numeric columns
                try {
                    List<String> numericCols = new ArrayList<>();
                    for (StructField field : schema.fields()) {
                        if (field.dataType().equals(DataTypes.IntegerType) || 
                            field.dataType().equals(DataTypes.DoubleType) ||
                            field.dataType().equals(DataTypes.LongType) ||
                            field.dataType().equals(DataTypes.FloatType)) {
                            numericCols.add(field.name());
                        }
                    }
                    
                    if (!numericCols.isEmpty()) {
                        logger.info("\n   Numeric Statistics:");
                        df.select(numericCols.toArray(new String[0])).describe().show();
                    }
                } catch (Exception e) {
                    logger.warn("Could not compute statistics: {}", e.getMessage());
                }
            }
        } catch (Exception e) {
            logger.error("Error generating summary for {}: {}", name, e.getMessage());
        }
    }
    
    /**
     * Setup checkpoint directory for streaming applications.
     * 
     * @param spark SparkSession
     * @param config SparkConfig
     * @param appName Application name for unique checkpoint directory
     */
    public static void setupKafkaCheckpoint(SparkSession spark, SparkConfig config, String appName) {
        String checkpointPath = config.getCheckpointDir() + "/" + appName;
        
        try {
            java.nio.file.Files.createDirectories(java.nio.file.Paths.get(checkpointPath));
            logger.info("✅ Checkpoint directory: {}", checkpointPath);
        } catch (Exception e) {
            logger.warn("⚠️  Could not create checkpoint directory: {}", e.getMessage());
            logger.warn("   Using default: /tmp/spark-checkpoints/{}", appName);
        }
    }
    
    /**
     * Parse command line arguments for Spark applications.
     * 
     * @param args Command line arguments
     * @return Map of parsed arguments with defaults
     */
    public static Map<String, String> parseArgs(String[] args) {
        Map<String, String> parsedArgs = new HashMap<>();
        
        // Set defaults
        parsedArgs.put("master", DEFAULT_MASTER);
        parsedArgs.put("app-name", DEFAULT_APP_NAME);
        parsedArgs.put("executor-memory", "2g");
        parsedArgs.put("executor-cores", "2");
        parsedArgs.put("records", "10000");
        parsedArgs.put("output-path", "/tmp/spark-output");
        parsedArgs.put("kafka-topic", "spark-demo");
        parsedArgs.put("duration", "60");
        parsedArgs.put("verbose", "false");
        
        // Parse command line arguments
        for (int i = 0; i < args.length; i++) {
            if (args[i].startsWith("--") && i + 1 < args.length) {
                String key = args[i].substring(2);
                String value = args[i + 1];
                parsedArgs.put(key, value);
                i++; // Skip the value in next iteration
            }
        }
        
        return parsedArgs;
    }
    
    /**
     * Print usage information for command line arguments.
     */
    public static void printUsage(String className) {
        System.out.println("Usage: " + className + " [OPTIONS]");
        System.out.println("Options:");
        System.out.println("  --master URL              Spark master URL (default: " + DEFAULT_MASTER + ")");
        System.out.println("  --app-name NAME           Spark application name (default: " + DEFAULT_APP_NAME + ")");
        System.out.println("  --executor-memory SIZE    Executor memory (default: 2g)");
        System.out.println("  --executor-cores NUM      Executor cores (default: 2)");
        System.out.println("  --records NUM             Number of records to process (default: 10000)");
        System.out.println("  --output-path PATH        Output path for results (default: /tmp/spark-output)");
        System.out.println("  --kafka-topic TOPIC       Kafka topic for streaming (default: spark-demo)");
        System.out.println("  --duration SECONDS        Duration for streaming jobs (default: 60)");
        System.out.println("  --verbose                 Enable verbose logging");
        System.out.println("  --help                    Show this help message");
    }
    
    /**
     * Properly clean up Spark session.
     * 
     * @param spark SparkSession to clean up
     */
    public static void cleanupSession(SparkSession spark) {
        try {
            logger.info("\n🔄 Cleaning up Spark session...");
            spark.stop();
            logger.info("✅ Spark session stopped");
        } catch (Exception e) {
            logger.warn("⚠️  Error during cleanup: {}", e.getMessage());
        }
    }
    
    /**
     * Create a SparkConfig from parsed command line arguments.
     * 
     * @param args Parsed command line arguments
     * @return SparkConfig object
     */
    public static SparkConfig createConfigFromArgs(Map<String, String> args) {
        SparkConfig config = new SparkConfig();
        config.setAppName(args.get("app-name"));
        config.setMaster(args.get("master"));
        config.setExecutorMemory(args.get("executor-memory"));
        config.setExecutorCores(args.get("executor-cores"));
        return config;
    }
    
    /**
     * Execute error handling wrapper for common operations.
     * 
     * @param operation The operation to execute
     * @param operationName Name of the operation for logging
     */
    public static void executeWithErrorHandling(Runnable operation, String operationName) {
        try {
            operation.run();
        } catch (Exception e) {
            logger.error("\n❌ Error in {}: {}", operationName, e.getMessage());
            logger.error("   Error type: {}", e.getClass().getSimpleName());
            e.printStackTrace();
            System.exit(1);
        }
    }
    
    /**
     * Get current timestamp as formatted string.
     * 
     * @return Formatted timestamp string
     */
    public static String getCurrentTimestamp() {
        return LocalDateTime.now().format(DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss"));
    }
    
    /**
     * Test the common utilities (main method for testing).
     */
    public static void main(String[] args) {
        System.out.println("🧪 Testing Spark common utilities...");
        
        SparkConfig config = new SparkConfig();
        config.setAppName("SparkCommonTest");
        
        SparkSession spark = createSparkSession(config);
        
        try {
            // Generate and show sample data
            Dataset<Row> df = generateSampleData(spark, 100);
            printDatasetSummary(df, "Sample Data", true);
            
            monitorJobProgress(spark, "Common Utilities Test");
            
            System.out.println("✅ Spark common utilities test completed");
            
        } finally {
            cleanupSession(spark);
        }
    }
}
